types/payload.go

```diff
@@ -0,0 +1,235 @@
+package types
+
+import (
+	"encoding/json"
+	"iter"
+	"maps"
+
+	jsoniter "github.com/json-iterator/go"
+	"github.com/vmihailenco/msgpack/v5"
+)
+
+type Payload struct {
+	// A serialized messagepack map used to source fields.
+	msgpMap *MsgpPayloadMap
+
+	// Deserialized fields, either from the internal msgpMap, or set externally.
+	memoizedFields map[string]any
+}
+
+func NewPayload(data map[string]any) Payload {
+	return Payload{
+		memoizedFields: data,
+	}
+}
+
+func (p *Payload) UnmarshalMsgpack(data []byte) error {
+	p.msgpMap = &MsgpPayloadMap{rawData: data}
+	p.memoizedFields = make(map[string]any)
+	return nil
+}
+
+func (p *Payload) UnmarshalJSON(data []byte) error {
+	var fields map[string]any
+	if err := jsoniter.Unmarshal(data, &fields); err != nil {
+		return err
+	}
+	p.memoizedFields = fields
+	return nil
+}
+
+// Extracts all of the listed fields from the internal msgp buffer in a single
+// pass, for efficient random access later.
+func (p *Payload) MemoizeFields(keys ...string) {
+	if p.msgpMap == nil {
+		return
+	}
+
+	if p.memoizedFields == nil {
+		p.memoizedFields = make(map[string]any, len(keys))
+	}
+
+	keysToFind := make(map[string]struct{}, len(keys))
+	for _, key := range keys {
+		if _, ok := p.memoizedFields[key]; !ok {
+			keysToFind[key] = struct{}{}
+		}
+	}
+	if len(keysToFind) == 0 {
+		return
+	}
+
+	iter, err := p.msgpMap.Iterate()
+	if err != nil {
+		return
+	}
+
+	var keysFound int
+	for keysFound < len(keysToFind) {
+		keyBytes, _, err := iter.NextKey()
+		if err != nil {
+			break
+		}
+
+		// Note we deliberately don't put string(keyBytes) in a variable here,
+		// because doing so will move it to the heap on every iteration.
+		// Keeping the string cast inline like this allows us to avoid the heap
+		// unless we're actually going to memoize the field.
+		if _, ok := keysToFind[string(keyBytes)]; ok {
+			value, err := iter.ValueAny()
+			if err != nil {
+				break
+			}
+			key := string(keyBytes)
+			p.memoizedFields[key] = value
+			keysFound++
+		}
+	}
+}
+
+func (p *Payload) Get(key string) any {
```

Comment:
I think returning `(any, bool)` like we used to get from a map would make using this func easier. There are many occasions where we try to get a value then immediately do `value != nil`.

Reply:
I think that depends on whether we actually want `nil` values or not. In this PR I assumed we actually don't want them (honeycomb will discard them on ingest) so this essentially pretends they don't exist. I probably should have noted that as a behavior change in the PR description.

Switching to the two-return form of Get() implies that you can get back a `nil` value and should do something with it, but the code here doesn't seem to really consider that. So, do you have an opinion there? Is there some reason why we want to distinguish "value is null" from "value does not exist at all"? If they're semantically the same I'd rather not have both states. (And returning `exists = false` any time the value is `nil` departs from the map semantic which I think is worse.)

Reply:
@VinozzZ has found that we include nil values in the trace key, meaning they have to be allowed to exist in the payload, so I'm going to switch to the 2-return form after all.